pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchClient.java

```diff
@@ -50,12 +51,17 @@ public ManagedChannel getChannel() {
   }
 
   public void submit(Worker.QueryRequest request, QueryServerInstance virtualServer, Deadline deadline,
-      Consumer<AsyncQueryDispatchResponse> callback) {
-    _dispatchStub.withDeadline(deadline).submit(request, new DispatchObserver(virtualServer, callback));
+      Consumer<AsyncResponse<Worker.QueryResponse>> callback) {
+    _dispatchStub.withDeadline(deadline).submit(request, new LastValueDispatchObserver<>(virtualServer, callback));
   }
 
   public void cancel(long requestId) {
     Worker.CancelRequest cancelRequest = Worker.CancelRequest.newBuilder().setRequestId(requestId).build();
     _dispatchStub.cancel(cancelRequest, NO_OP_CANCEL_STREAM_OBSERVER);
   }
+
+  public void explain(Worker.QueryRequest request, QueryServerInstance virtualServer, Deadline deadline,
+      Consumer<AsyncResponse<List<Worker.ExplainResponse>>> callback) {
+    _dispatchStub.withDeadline(deadline).explain(request, new AllValuesDispatchObserver<>(virtualServer, callback));
```

Comment:
I don't know why we only keep the last message in the submit case, but you can see the older implementation is using [DispatchObserver](https://github.com/apache/pinot/blob/master/pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/DispatchObserver.java) and verify that it is keeping only the latest `QueryResponse` received in `onNext`, which is what `LastValueDispatcherObserver` does.

TBH the old behavior is correct. You can see that `org.apache.pinot.query.service.server.QueryServer.submit` does only call `onNext` once, so it is ok to keep the last value. What I don't get is why the proto file is defined as:
```
  rpc Submit(ServerRequest) returns (stream ServerResponse);
```
instead of
```
  rpc Submit(ServerRequest) returns (ServerResponse);
```

Anyway, if we change the server implementation we will need to use the new `AllValuesDispatcherObserver` and change the code a bit to work with a list instead of a single element.

Reply:
> What I don't get is why the proto file is defined as:
>
>  ```rpc Submit(ServerRequest) returns (stream ServerResponse);```
> instead of
>
> ```  rpc Submit(ServerRequest) returns (ServerResponse);```

That's the proto definition for [GrpcQueryServer](https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-core/src/main/java/org/apache/pinot/core/transport/grpc/GrpcQueryServer.java#L65) right (which is for v1 streaming queries FWICT)? The proto definition for the multi-stage engine's [QueryServer](https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java#L50) `Submit` RPC is - https://github.com/apache/pinot/blob/de577bc457b580e89ecb4e82076ce09f209bca18/pinot-common/src/main/proto/worker.proto#L26

So I guess this makes sense now since the new `Explain` RPC returns a `stream ExplainResponse` and the implementation does call `onNext` multiple times.

Reply:
You are right, the Submit endpoint is not defined as stream, so it makes sense to only keep the last (as only 1 is expected).

BTW I don't think submit is used for V1. Remember we have two classes called QueryServer. `org.apache.pinot.query.service.server.QueryServer` is the one used in V2 and `org.apache.pinot.core.transport.QueryServer` is the one used in V1. Then we have ` org.apache.pinot.core.transport.grpc.GrpcQueryServer`. The later is used to execute V1 queries using GPRC and implements the protocol defined in `server.proto`.